unified-runtime/source/adapters/offload/device.cpp

```diff
@@ -207,14 +208,32 @@ UR_APIEXPORT ur_result_t UR_APICALL urDeviceSelectBinary(
   return UR_RESULT_ERROR_INVALID_BINARY;
 }
 
-UR_APIEXPORT ur_result_t UR_APICALL
-urDeviceGetNativeHandle(ur_device_handle_t, ur_native_handle_t *) {
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+UR_APIEXPORT ur_result_t UR_APICALL urDeviceGetNativeHandle(
+    ur_device_handle_t UrDevice, ur_native_handle_t *Handle) {
+  *Handle = reinterpret_cast<ur_native_handle_t>(UrDevice->OffloadDevice);
+  return UR_RESULT_SUCCESS;
 }
 
 UR_APIEXPORT ur_result_t UR_APICALL urDeviceCreateWithNativeHandle(
-    ur_native_handle_t, ur_adapter_handle_t,
-    const ur_device_native_properties_t *, ur_device_handle_t *) {
+    ur_native_handle_t hNativeDevice, ur_adapter_handle_t hAdapter,
+    const ur_device_native_properties_t *, ur_device_handle_t *phDevice) {
+  ol_device_handle_t OlDevice =
+      reinterpret_cast<ol_device_handle_t>(hNativeDevice);
+
+  // Currently, all devices are found at initialization, there is no way to
+  // create sub devices yet
+  for (auto &P : hAdapter->Platforms) {
+    auto Found =
+        std::find_if(P->Devices.begin(), P->Devices.end(),
+                     [&](std::unique_ptr<ur_device_handle_t_> &PDevice) {
+                       return PDevice->OffloadDevice == OlDevice;
+                     });
+    if (Found != P->Devices.end()) {
+      *phDevice = Found->get();
+      return UR_RESULT_SUCCESS;
+    }
+  }
+
   return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
```

Comment:
Probably better to return `UR_RESULT_ERROR_INVALID_VALUE` or something if the native handle does match a known device.

Reply:
The spec says:
```
* UR_RESULT_ERROR_UNSUPPORTED_FEATURE
    If the adapter has no underlying equivalent handle.
```

Which I think applies here. `INVALID_VALUE` isn't also a legal return type according to the docs.